inline_translated.c

```diff
@@ -0,0 +1,554 @@
+register cpu_t * pcpu asm("r15");
+
+/* Area for generated code. It is put into the .text section to be reachable
+   from the rest of the code (relative branch to fit in 32 bits) */
+/* For explanation of '#' character,
+   see https://gcc.gnu.org/ml/gcc-help/2010-09/msg00088.html */
+char gen_code[JIT_CODE_SIZE] __attribute__ ((section (".text#")))
+                             __attribute__ ((aligned(4096)));
+
+/* TODO:a global - not good. Should be moved into cpu state or somewhere else */
+static long long steplimit = LLONG_MAX;
+
+static inline decode_t decode_at_address(const Instr_t* prog, uint32_t addr) {
+    assert(addr < PROGRAM_SIZE);
+    decode_t result = {0};
+    Instr_t raw_instr = prog[addr];
+    result.opcode = raw_instr;
+    switch (raw_instr) {
+    case Instr_Nop:
+    case Instr_Halt:
+    case Instr_Print:
+    case Instr_Swap:
+    case Instr_Dup:
+    case Instr_Inc:
+    case Instr_Add:
+    case Instr_Sub:
+    case Instr_Mul:
+    case Instr_Rand:
+    case Instr_Dec:
+    case Instr_Drop:
+    case Instr_Over:
+    case Instr_Mod:
+    case Instr_And:
+    case Instr_Or:
+    case Instr_Xor:
+    case Instr_SHL:
+    case Instr_SHR:
+    case Instr_Rot:
+    case Instr_SQRT:
+    case Instr_Pick:
+        result.length = 1;
+        break;
+    case Instr_Push:
+    case Instr_JNE:
+    case Instr_JE:
+    case Instr_Jump:
+        result.length = 2;
+        assert(addr+1 < PROGRAM_SIZE);
+        result.immediate = (int32_t)prog[addr+1];
+        break;
+    case Instr_Break:
+    default: /* Undefined instructions equal to Break */
+        result.length = 1;
+        result.opcode = Instr_Break;
+        break;
+    }
+    return result;
+}
+
+/* Supporting functions for in-lined service routines */
+
+static void enter_generated_code(void* addr) {
+    __asm__ __volatile__ ( "jmp *%0"::"r"(addr):);
+}
+
+static void exit_generated_code() {
+    longjmp(return_buf, 1);
+}
+
+static inline void push(cpu_t *pcpu, uint32_t v) {
+    assert(pcpu);
+    if (pcpu->sp >= STACK_CAPACITY-1) {
+        printf("Stack overflow\n");
+        pcpu->state = Cpu_Break;
+        exit_generated_code();
+    }
+    pcpu->stack[++pcpu->sp] = v;
+}
+
+static void inline_translate_program(const Instr_t *prog,
+                           char *out_code, void **entrypoints, int len) {
+    assert(prog);
+    assert(out_code);
+    assert(entrypoints);
+
+    /* An IA-32 instruction "CALL rel32" is used as a trampoline to invoke
+       service routines. A template for it is "call .+0x00000005" */
+    const char call_template_code[] = { 0xe8, 0x00, 0x00, 0x00, 0x00 };
+    const int call_template_size = sizeof(call_template_code);
+
+    int i = 0; /* Address of current guest instruction */
+    char* cur = out_code; /* Where to put new code */
+
+    while (i < len) {
+        decode_t decoded = decode_at_address(prog, i);
+        entrypoints[i] = (void*) cur;
+
+        int addr_exit1 = 0;

```

Comment:
An explanation of what these variables are is actually required. Otherwise it is impossible to understand what happens next.